docs/DeveloperGuide.md

```diff
@@ -1,38 +1,584 @@
+
+* **Alternative 1 (current choice):** Maintain an individual array list for found income and expenditure
+    * Pros: Easy and fast access of found results, no additional logic needed to separate income from expenditure
+    * Cons: Additional memory management, needs to manage 2 array list.
+
+* **Alternative 2 :** Maintain a single array list for both found income and expenditure
+    * Pros: Easy to implement and less memory usage.
+    * Cons: Additional logic check is needed to print the income and expenditure in a well formatted way
+
+### [Proposed] Listing Categorised Expenditures Feature
+
+#### Proposed Implementation
+
+The proposed listing categorised expenditure feature is facilitated by `Categories`,
+`Expenditure`, `ExpenditureList`  and `Family`. The `Categories` is an enumeration of keys that is used as the
+expenditure categories. The `Expenditure` will have an additional integer attribute that acts as an index to a category.
+Additionally, the following operations are implemented:
+
+* `Categories#getLabel(index)` -- Returns the name of the category with this index.
+* `Expenditure#getCategory()` -- Returns the category index for this expense.
+* `ExpenditureList#getExpenseOfCategory(index)` -- Returns a list of expenditures having category index matching the
+  index argument.
+* `Family#listExpenseOfCategory(index)` -- Lists all expenses from the category index.
+
+Below is an example usage scenario and how the expenses of a category will be printed.
+
+Step 1. Given that the application already has existing data and there are two people being tracked, Alice and Bob, and
+only Alice's expenses were added and categorised. Suppose that Alice is the main user and Bob is her father, then Alice
+would belong to the current generation and Bob would belong to the parent generation. In this case `Family` would be
+initialised with two generations being tracked - parents and myGen.
+
+![ListCategorisedExpense0](images/ListCategorisedExpense0.png)
+
+Step 2. The user executes `listcat /c 1` command to list all expenses in category `1`. The `listcat`
+command will be parsed and calls `Family#listExpenseOfCategory(1)` which would instantiate a temporary array list for
+storing the results of the upcoming search.
+
+![ListCategorisedExpense1](images/ListCategorisedExpense1.png)
+
+Step 3. After the temporary array list has been created, the generations being tracked will be iterated for `Person`
+objects. The `expenditureList` for a person would be retrieved during that person's iteration
+and `ExpenditureList#getExpenseOfCategory(1)` will be called as `expenditureList`
+extends `ExpenditureList`. This method then iterates through the list and calls
+`Expenditure#getCategory()` on each expenditure, collecting and returning the expenditure if its category matches the
+given index. The returned expenditures are then appended to the temporary array list.
+
+![ListCategorisedExpense2](images/ListCategorisedExpense2.png)
```

Comment:
Maybe can consider to show just the important parts of this diagrams? It seems that the point is to illustrate the new state of the tempList so maybe the other details not too important?
![image](https://user-images.githubusercontent.com/58723008/161183813-05f8385a-b774-4434-ab58-d199e2d7943f.png)